File: frontend/src/types/mcp.ts
/**
 * MCP Server and Timeline types for transparent demo interface.
 */

// ===========================================
// MCP Server Types
// ===========================================

export type McpServerStatus = 'connected' | 'disconnected' | 'error' | 'loading';

export interface McpServer {
  name: string;
  transport: string;
  connected: boolean;
  enabled: boolean;
  toolsCount: number;
  tools: string[];
  lastPingMs?: number;
}

export interface McpServerConfig {
  name: string;
  enabled: boolean;
}

// ===========================================
// Timeline Types
// ===========================================

export type ToolCallStatus = 'pending' | 'running' | 'success' | 'error';

export interface TimelineToolCall {
  id: string;
  messageId: string;
  conversationId: string;
  toolName: string;
  serverName: string;
  arguments: Record<string, unknown>;
  status: ToolCallStatus;
  startTime: Date;
  endTime?: Date;
  durationMs?: number;
  resultPreview?: string;
  success?: boolean;
}

export interface TimelineEvent {
  id: string;
  type: 'tool_call_start' | 'tool_call_end' | 'content_start' | 'content_end';
  timestamp: Date;
  data: TimelineToolCall | { messageId: string };
}

// ===========================================
// Model Types
// ===========================================

export type ModelStatus = 'available' | 'loading' | 'pulling' | 'error';

export interface Model {
  name: string;
  size: number;
  modifiedAt: string | null;
  digest: string;
}

export interface ModelPullProgress {
  status: string;
  digest?: string;
  total?: number;
  completed?: number;
}

// ===========================================
// Health Types
// ===========================================

export type ComponentHealthStatus = 'healthy' | 'unhealthy' | 'degraded';

export interface ComponentStatus {
  status: ComponentHealthStatus;
  details: Record<string, unknown>;
}

export interface HealthResponse {
  status: ComponentHealthStatus;
  components: Record<string, ComponentStatus>;
  version: string;
}

File: backend/app/api/schemas/mcp.py
"""MCP-related schemas."""

from typing import Any, Dict, List, Optional
from pydantic import BaseModel


class ToolInfo(BaseModel):
    """Information about an available tool."""

    name: str
    description: str
    input_schema: Dict[str, Any]
    server_name: str


class ToolExecution(BaseModel):
    """Details of a tool execution."""

    name: str
    arguments: Dict[str, Any]
    result_preview: str
    result_full: Optional[Any] = None
    duration_ms: int
    success: bool
    server_name: str


class MCPServerInfo(BaseModel):
    """Information about an MCP server."""

    name: str
    transport: str
    connected: bool
    enabled: bool
    tools_count: int
    tools: List[str] = []
    last_ping_ms: Optional[int] = None


class MCPServerToggleRequest(BaseModel):
    """Request to toggle MCP server status."""

    enabled: bool
